soc: intel: cavs: add secondary core context save support

RanderWang opened the pull request:
The secondary core context save is already supported by ace platform but missed on cavs platform. Now add it for cavs.

Validated on TGL platforms

RanderWang committed:
soc: intel_adsp/cavs: add secondary dsp context save support

Save secondary dsp context when it is powered off in idle thread
and restore it when the secondary dsp is powered up. The algorithm
is aligned with ace platform.

Tested on a tgl platform.

Signed-off-by: Rander Wang <rander.wang@intel.com>

jxstelter commented on soc/xtensa/intel_adsp/common/include/cpu_init.h:
Set of registers needed to store for core context depends on core definition.  It is identical for CAVS and MTL/LNL from ACE, but this could change for other cores. Therefore moving it to common could ends up in future with some core dependent #ifdefs() sections.

RanderWang replied:
sure, I can image it. Let's move it to a new header file for current dsp

andyross commented on soc/xtensa/intel_adsp/common/multiprocessing.c:
```diff
@@ -96,6 +96,24 @@ __asm__(".section .text.z_soc_mp_asm_entry, \"x\" \n\t"
 
+__asm__(".align 4\n\t"
+	".global dsp_restore_vector\n\t"
+	"dsp_restore_vector:\n\t"
```
Design note:

This is like the fifth or possibly sixth mostly-cut-and-paste copy of this bootstrap code.  It's an inline inside C code only because the first version (which I did for cavs15, I think?) was ~5 instructions and didn't need to be anything fancy.  But at this point it's grown well beyond that, and every version has its own hackery (like the IMHO senseless "compute a runtime array index based on processor ID to get our stack pointer").

What this really wants to be is some kind of build-time gadget (maybe a C macro, maybe a gas/asm macro, maybe a python script, maybe even a set of .S files generated from cmake) that does something like "*Please emit a bootstrap sequence that can be jumped to via this symbol name, that sets PS to this value, that jumps to this symbol after setup, and that uses this stack pointer*"

andyross commented on soc/xtensa/intel_adsp/common/multiprocessing.c:
```diff
@@ -96,6 +96,24 @@ __asm__(".section .text.z_soc_mp_asm_entry, \"x\" \n\t"
+	"dsp_restore_vector:\n\t"
+	"  movi  a0, 0\n\t"
+	"  movi  a1, 1\n\t"
+	"  movi  a2, 0x40020\n\t"/* PS_UM|PS_WOE */
```
Note that once call0 lands, we need a way to select whether or not WOE is set (though I guess in a pinch we could do that in the C entry function -- this is a benign noop as long as code is call0-ABI-pure, as nothing will try to issue an ENTRY or RETW instruction).

tmleman replied:
We are here using xtensa feature with register windows. After core context restore instruction `retw` will result with jump back to the place that lead us tlow power state (pm subsystem in idle thread).

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -119,37 +105,13 @@ struct lpsram_header {
-	core_desc[core_id].vecbase = XTENSA_RSR("VECBASE");
-	core_desc[core_id].excsave2 = XTENSA_RSR("EXCSAVE2");
-	core_desc[core_id].excsave3 = XTENSA_RSR("EXCSAVE3");
-	core_desc[core_id].thread_ptr = XTENSA_RUR("THREADPTR");
```
I understand this is just code motion, but frankly I think these are being stored in the wrong place.

* VECBASE is known architecturally, we should use the same static symbol address we use at CPU initialization time.

* THREADPTR is just the current thread struct, which is already stored in memory (`_kernel.cpus[_current_cpu].current`)

* EXCSAVEn registers are owned by the arch layer and are used through the abstraction layer in zsr.h, they shouldn't be touched outside arch/xtensa for any reason.  Why are we doing this?

andyross replied:
Coming back having looked more carefully: my guess is that we're saving EXCSAVE2/3 because we need to stash the CPU pointer register, which I don't see being restored any other way.  Yeah, don't do it this way, use the ZSR_CPU alias, that's what it's for.

andyross replied:
Also I was a little mixed up: threadptr is not the thread struct itself, it's the tls field of the thread struct.

And... IIRC it's not in use anywhere on SOF platforms because of the lack of support for TLS in the cadence toolchains?  Is this another spot where we're using a register we shouldn't?

RanderWang replied:
Thanks for your valuable advice. I will spend time to digest it and try to improve it.

lgirdwood replied:
@andyross the HW state being saved here is also Intel implementation specific hence being in SoC specific code.

andyross commented on soc/xtensa/intel_adsp/common/include/cpu_init.h:
```diff
@@ -101,4 +101,42 @@ static ALWAYS_INLINE void cpu_early_init(void)
+	uint32_t excsave3;
+	uint32_t thread_ptr;
+	uint32_t intenable;
+	uint32_t bctl;
```
I can't see anything actually touching the intenable or bctl fields?  Did that get dropped or is it not part of the code motion?

RanderWang replied:
it is used by "core_desc[cpu].intenable = XTENSA_RSR("INTENABLE"); " and "	core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;"

andyross commented on soc/xtensa/intel_adsp/common/include/cpu_init.h:
```diff
@@ -101,4 +101,42 @@ static ALWAYS_INLINE void cpu_early_init(void)
+	core_desc[core_id].excsave3 = XTENSA_RSR("EXCSAVE3");
+	core_desc[core_id].thread_ptr = XTENSA_RUR("THREADPTR");
+	__asm__ volatile("mov %0, a0" : "=r"(core_desc[core_id].a0));
+	__asm__ volatile("mov %0, a1" : "=r"(core_desc[core_id].a1));
```
Yikes.  You're stashing the return address and stack pointer here, in a random function in the middle of some C processing that is going to **continue**, then at restore you push these values back into the registers and not the ones that were actually present in the code doing the suspend!?

I mean... I see that this is ALWAYS_INLINE, which means that this code isn't clearly 100% wrong, but... yikes.  All it takes is for someone to helpfully move this into a sub-function and everything will blow up in a way that almost no one will be able to successfully debug.

Can we please move the GPR saving to the point of suspend so that you restore to exactly the same state that suspended?

(Alternatively: have the suspend happen out of a known-restartable context like the idle thread, and just run a fixed function at startup that doesn't need to restore GPR state)

RanderWang replied:
@andyross  save_core_context is already in idle thread when powering off dsp core.
````
 idle thread
    pm_system_suspend
       pm_state_set ( SOFT_OFF)
          save_dsp_context
````
After restore_dsp_context, the idle thread will get out of pm_state_set and go on executing. Do I miss something ? thanks!

tmleman replied:
This is not a random place and its not a function you can call from random place. It's executed at the end of cpu shutdown process and and we don't have much flexibility here when it comes to moving this code to other places.

andyross reviewed (changes requested):
A few notes.  I get that this is just code motion and that most of this is already in the tree.  But I'm going to be sticky over the EXCSAVE usage, because that looks pretty wrong and should be easy to fix if it's there for the reason I think.

The other notes are of the "really we need to fix this but maybe not now" form.

tmleman reviewed (commented):
LGTM, I just wish there were run tests first on TGL, MTL and LNL before merge.

I think @andyross  comments should be addressed in another PR. The author's intention was to move the code to a place so it could be used elsewhere. Since this code takes care of the power flows, each change must be thoroughly tested.

RanderWang commented:
@andyross  @tmleman  Update the PR according to your comments.  Thanks!

tmleman reviewed (commented):
Typos in commit description:

> Save secondary dsp **contest** when it is powered **of** in idle thread

RanderWang commented:
> Typos in commit description:
> 
> > Save secondary dsp **contest** when it is powered **of** in idle thread

thanks, updated.

tmleman commented:
> > Typos in commit description:
> > > Save secondary dsp **contest** when it is powered **of** in idle thread
> 
> thanks, updated.

is powered **off**?

RanderWang commented:
> > > Typos in commit description:
> > > > Save secondary dsp **contest** when it is powered **of** in idle thread
> > 
> > 
> > thanks, updated.
> 
> is powered **off**?

o, thanks! updated.

RanderWang commented:
@tmleman  update, thanks for review!

RanderWang commented:
@andyross  Could you help to review it ?  Thanks!

andyross commented on soc/xtensa/intel_adsp/cavs/power.c:
```diff
@@ -76,6 +79,51 @@ static inline void __sparse_cache *uncache_to_cache(void *address)
+{
+	uint32_t core_id = arch_proc_id();
+
+	core_desc[core_id].excsave2 = XTENSA_RSR(ZSR_CPU_STR);
```
This is correct, but calling the field "excsave2" is correct, but semantically fragile.  All it takes is for someone to come along and change the register ordering in gen_zsr.py and we'll select a different register to store in that field.  Would personally want to see this called "cpu" or "cpu_ptr" or whatever, but that's just a nitpick.

andyross reviewed (approved):
Rander had to drag me back here.  My big concerns were all addressed.

I still think that writing SoC-specific context saving code is kinda silly given that the OS can already do this for you, and is going to result in needless breakage and churn in the future as the OS arch layer changes.

RanderWang commented:
@nashif  Could you help to review this PR ?  Thanks!
